src/cachingreaderworker.cpp

```diff
@@ -1,122 +1,66 @@
 #include <QtDebug>
 #include <QFileInfo>
+#include <QMutexLocker>
 
 #include "controlobject.h"
 #include "controlobjectthread.h"
 
 #include "cachingreaderworker.h"
-#include "trackinfoobject.h"
-#include "sampleutil.h"
 #include "soundsourceproxy.h"
 #include "util/compatibility.h"
 #include "util/event.h"
-#include "util/math.h"
-
-
-// One chunk should contain 1/2 - 1/4th of a second of audio.
-// 8192 frames contain about 170 ms of audio at 48 kHz, which
-// is well above (hopefully) the latencies people are seeing.
-// At 10 ms latency one chunk is enough for 17 callbacks.
-// Additionally the chunk size should be a power of 2 for
-// easier memory alignment.
-// TODO(XXX): The optimum value of the "constant" kFramesPerChunk
-// depends on the properties of the AudioSource as the remarks
-// above suggest!
-const SINT CachingReaderWorker::kChunkChannels = Mixxx::AudioSource::kChannelCountStereo;
-const SINT CachingReaderWorker::kFramesPerChunk = 8192; // ~ 170 ms at 48 kHz
-const SINT CachingReaderWorker::kSamplesPerChunk = kFramesPerChunk * kChunkChannels;
-
-CachingReaderWorker::CachingReaderWorker(QString group,
-        FIFO<ChunkReadRequest>* pChunkReadRequestFIFO,
+
+
+CachingReaderWorker::CachingReaderWorker(
+        QString group,
+        FIFO<CachingReaderChunkReadRequest>* pChunkReadRequestFIFO,
         FIFO<ReaderStatusUpdate>* pReaderStatusFIFO)
         : m_group(group),
           m_tag(QString("CachingReaderWorker %1").arg(m_group)),
           m_pChunkReadRequestFIFO(pChunkReadRequestFIFO),
           m_pReaderStatusFIFO(pReaderStatusFIFO),
+          m_maxReadableFrameIndex(Mixxx::AudioSource::getMinFrameIndex()),
           m_stop(0) {
 }
 
 CachingReaderWorker::~CachingReaderWorker() {
 }
 
-void CachingReaderWorker::processChunkReadRequest(
-        ChunkReadRequest* request,
+void CachingReaderWorker::processCachingReaderChunkReadRequest(
+        CachingReaderChunkReadRequest* request,
         ReaderStatusUpdate* update) {
-    //qDebug() << "Processing ChunkReadRequest for" << chunk_number;
+    //qDebug() << "Processing CachingReaderChunkReadRequest for" << chunk_number;
 
-    // Initialize the output parameter
+    // Initialize the output parameters
+    update->maxReadableFrameIndex = m_maxReadableFrameIndex;
```

Comment:
I've added more comments to the worker that hopefully explain why we need 
two invocations of isReadable(), one before and one after reading data from 
the audio source to determine the result status of the read request.

On 07/27/2015 02:44 PM, Daniel Schürmann wrote:
>
> In src/cachingreaderworker.cpp 
> <https://github.com/mixxxdj/mixxx/pull/652#discussion_r35531074>:
>
> >
> > -    // Initialize the output parameter
> > +    // Initialize the output parameters
> > +    update->maxReadableFrameIndex = m_maxReadableFrameIndex;
>
> Since the m_maxReadableFrameIndex may be changed below, is it line here 
> corrector?
> maybe we can entirely ditch m_maxReadableFrameIndex and use always 
> directly the m_pAudioSource value.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/652/files#r35531074>.
>